modin/pandas/groupby.py

```diff
@@ -634,7 +622,9 @@ def corr(self, **kwargs):
         return self._default_to_pandas(lambda df: df.corr(**kwargs))
 
     def fillna(self, **kwargs):
-        result = self._apply_agg_function(lambda df: df.fillna(**kwargs))
+        result = self._apply_agg_function(
+            lambda df: df.fillna(**kwargs), overwrite_groupby_kwargs={"as_index": True}
```

Comment:
Why not just do the `kwargs.update({"as_index": True}) in the function instead of adding a new argument to `_apply_agg_function`?

Reply:
Because `kwargs` in this context is an argument for `fillna` function, not for groupby.

If you were speaking about `self._kwargs` then we should also revert this change after getting the result since we don't want to modify GroupBy object. Is it what you've proposed?

Reply:
Yes, that is what I mean, I see we should update the `self._kwargs` afterward to reset it. This looks like a regression in pandas, was this reported there?

Reply:
This is not regression, 'fillna` does not consider `as_index` parameter at all, so to bypass our `as_index=False` logic that incorrectly processes `fillna` index, it's better set `as_index=True`.

Why we haven't seen this problem before? At 1.2.0 version pandas fixed a bug, where `dropna=True` parameter doesn't drop `nan` groups.
```python
import pandas
import numpy as np

data = {"a": [np.nan, 2, np.nan, 2], "b": [1, 2, 3, np.nan]}
df = pandas.DataFrame(data)
res1 = df.groupby("a", as_index=False).fillna(method="ffill")
res2 = df.groupby("a", as_index=True).fillna(method="ffill")
assert res1.equals(res2)
print(res1)

# pandas 1.1.5 output:
#      b
# 0  NaN
# 1  2.0
# 2  NaN
# 3  2.0

# pandas 1.2.0+ output
#      b
# 1  2.0
# 3  2.0
```
The latest pandas version processes the` dropna=True` parameter correctly and drops `nan` groups, so the index is not monotonic range anymore (as our logic considers it when `as_index=False`).

Reply:
The problem with modifying and restoring class attribute is that if an exception occurs somewhere between modifying and restoring the value, then we will get a modified object:
<details><summary>Code example</summary>

```python
class Example:
    def __init__(self):
        self.attr = "initial value"
    
    def method(self):
        previous_value = self.attr
        self.attr = "modified value"
        self._private_method()
        self.attr = previous_value
    
    def _private_method(self):
        raise Exception

obj = Example()
print(obj.attr)
try:
    obj.method()
except:
    pass
print(obj.attr)

# Output:
# initial value
# modified value
```

</details>

Reply:
There is another way how we can modify groupby parameter for a single operation - we can create a new groupby object like we do in [`groupby.size()`](https://github.com/modin-project/modin/blob/8a50c4a147de94b10bf098b5f540d22c43c76268/modin/pandas/groupby.py#L492). But IMHO this makes code pretty heavy, and also triggers [unnecessary computations](https://github.com/modin-project/modin/blob/8a50c4a147de94b10bf098b5f540d22c43c76268/modin/pandas/groupby.py#L75-L92) in the GroupBy constructor.

Reply:
I think it is probably fine to create a new GroupBy object, the computation should be relatively small (`by` should not be very long).

Reply:
done